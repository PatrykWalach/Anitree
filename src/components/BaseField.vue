<template>
  <component :is="tag" v-model="input" v-bind="$attrs" @change="changeInput">
    <slot></slot>
  </component>
</template>

<script lang="ts">
const BaseDateField = () => import('./BaseDateField.vue')
import { VChipGroup, VSelect, VTextField, VTextarea } from 'vuetify/lib'

import { computed, createComponent, ref, watch } from '@vue/composition-api'

export interface Props {
  validators: ((v: string) => boolean)[]
  transformations: ((v: string) => string)[]
  tag: string
  beforeTransform: ((v: any) => any)[]
  afterTransform: ((v: any) => any)[]
  value: any
}

export default createComponent<Readonly<Props>>({
  components: {
    BaseDateField,
    VChipGroup,
    VSelect,
    VTextField,
    VTextarea
  },
  inheritAttrs: false,
  props: {
    afterTransform: { default: () => [], required: false, type: Array },
    beforeTransform: { default: () => [], required: false, type: Array },
    tag: {
      default: 'v-text-field',
      required: false,
      type: String
    },
    transformations: { default: () => [], required: false, type: Array },
    validators: { default: () => [], required: false, type: Array },
    value: {
      default: '',
      required: false,
      type: null
    }
  },
  setup(props, { emit }) {
    const input = ref('')
    const transform = (input: any, transformations: Function[]) =>
      transformations.reduce(
        (str, transformation) => transformation(str),
        input
      )

    watch(
      () => props.value,
      _value => {
        input.value = transform(_value, props.beforeTransform)
      }
    )

    const isValid = computed(() => {
      for (const validator of props.validators) {
        if (!validator(input.value)) {
          return false
        }
      }
      return true
    })

    const changeInput = (newValue: string) => {
      input.value = transform(newValue, props.transformations)
      if (isValid.value) {
        emit('input', transform(input.value, props.afterTransform))
      } else {
        input.value = transform(props.value, props.beforeTransform)
      }
    }

    return { changeInput, input, isValid }
  }
})
</script>
