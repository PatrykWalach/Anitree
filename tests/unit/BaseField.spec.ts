import { createLocalVue, shallowMount } from '@vue/test-utils'
import BaseField from '@/components/BaseField.vue'
import VueCompostionApi from '@vue/composition-api'
import Vuetify from 'vuetify'

describe('BaseField', () => {
  const localVue = createLocalVue()
  localVue.use(Vuetify)
  localVue.use(VueCompostionApi)
  test('matches snapshot', () => {
    const wrapper = shallowMount(BaseField, {
      localVue,
      propsData: {
        value: '',
      },
      stubs: {
        VTextField: '<input/>',
      },
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
