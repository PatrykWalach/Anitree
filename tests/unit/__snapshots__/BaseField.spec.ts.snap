// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseField matches snapshot 1`] = `
<input
  value=""
/>
`;
